VisualPinball.Unity/VisualPinball.Unity.Editor/Managers/ManagerWindow.cs

```diff
@@ -0,0 +1,288 @@
+using System.Collections.Generic;
+using System.Reflection;
+using UnityEditor;
+using UnityEditor.IMGUI.Controls;
+using UnityEngine;
+using VisualPinball.Unity.Extensions;
+using VisualPinball.Unity.VPT;
+using VisualPinball.Unity.VPT.Table;
+
+namespace VisualPinball.Unity.Editor.Managers
+{
+	/// <summary>
+	/// Base class for VPX-style "Manager" windows, such as the Material Manager
+	/// </summary>
+	/// <typeparam name="T">class of type IManagerListData that represents the data being edited</typeparam>
+	public abstract class ManagerWindow<T> : EditorWindow where T: class, IManagerListData
+	{
+		protected virtual string DataTypeName => "";
+
+		protected virtual void OnButtonBarGUI() { }
+		protected virtual void OnDataDetailGUI() { }
+		protected virtual void RenameExistingItem(T data, string desiredName) { }
+		protected virtual void CollectData(List<T> data) { }
+		protected virtual void OnDataChanged(string undoName, T data) { }
+		protected virtual void AddNewData(string undoName, string newName) { }
+		protected virtual void RemoveData(string undoName, T data) { }
+		protected virtual void CloneData(string undoName, string newName, T data) { }
+
+		protected TableBehavior _table;
+		protected T _selectedItem;
+
+		private List<T> _data = new List<T>();
+		private ManagerListView<T> _listView;
+		private TreeViewState _treeViewState;
+		private bool _renaming = false;
+		private string _renameBuffer = "";
+		[SerializeField] private string _forceSelectItemWithName;
+
+		protected void Reload()
+		{
+			_data.Clear();
+			CollectData(_data);
+			_listView.SetData(_data);
+		}
+
+		protected virtual void OnEnable()
+		{
+			// force gui draw when we perform an undo so we see the fields change back
+			Undo.undoRedoPerformed -= UndoPerformed;
+			Undo.undoRedoPerformed += UndoPerformed;
+
+			if (_treeViewState == null) {
+				_treeViewState = new TreeViewState();
+			}
+
+			FindTable();
+		}
+
+		protected virtual void OnHierarchyChange()
+		{
+			// if we don't have a table, look for one when stuff in the scene changes
+			if (_table == null) {
+				FindTable();
+			}
+		}
+
+		protected virtual void OnGUI()
+		{
+			// if the table went away, clear the selected material and list data
+			if (_table == null) {
+				_selectedItem = null;
+				_listView?.SetData(null);
+			}
+
+			if (!string.IsNullOrEmpty(_forceSelectItemWithName)) {
+				_listView.SelectItemWithName(_forceSelectItemWithName);
+				_forceSelectItemWithName = null;
+			}
+
+			EditorGUILayout.BeginHorizontal();
+			if (IsImplemented("AddNewData") && GUILayout.Button("Add", GUILayout.ExpandWidth(false))) {
```

Comment:
I learned about `[GUI.enabled](https://docs.unity3d.com/ScriptReference/GUI-enabled.html)` yesterday - maybe disable the button when there's no implementation?

Reply:
Personally I would interpret a disabled button as something that *could* be used, just not for the selected item.

Reply:
You're right, and now I'm reading it again, it doesn't actually show the button if there's no implementation (I thought otherwise).